### ISSUE TITLE
Declaring classes as final

### ISSUE THREAD

**Issue by leofeyer** [contao/contao#924]
**Why?**

https://ocramius.github.io/blog/when-to-declare-classes-final/

**If and how?**

That remains to be discussed. 😄 

**Related**

https://github.com/contao/docs/pull/138

**Comment by aschempp** [comment]
Honestly, I don't really agree with this. For the mentioned case, we are using `private` methods in our classes. Making classes `final` does not bring any advantage, because an inheriting class can still only call the public methods. If a method is `protected`, then it is **explicitly** mean to be used in child classes. Also, `final` conflicts a lot with unit tests (as far as I know).

**Comment by m-vo** [comment]
I agree with the reasoning in the linked article. It's imho good practice - not only in PHP ("inheritance breaks encapsulation").

There might be a problem with mocking or using proxies like [Ocramius/ProxyManager](https://github.com/Ocramius/ProxyManager) (how ironically). On the other hand there should probably be **interfaces** in that case (also for mocking) rendering this problem obsolete.

But again, that's also in the article:
>  Final classes only work effectively under following assumptions:
>
>    There is an abstraction (interface) that the final class implements
>   All of the public API of the final class is part of that interface

Maybe also have a look at these:
* https://stackoverflow.com/questions/11730352/circumstances-to-make-a-class-final
* https://stackoverflow.com/questions/9344935/inheritance-breaking-encapsulation

**Comment by leofeyer** [comment]
I also agree with the reasoning in the article.

**Comment by leofeyer** [comment]
We should at least make the `@internal` classes final in Contao 6.

**Comment by m-vo** [comment]
Yes, but only if they implement an interface or do not require mocking. 🙃 Otherwise I'd vote for using the `@final` annotation.
